File: miniprojet.r
# ======================
# 1. INSTALLATION ET CHARGEMENT DES LIBRAIRIES
# ======================
install.packages(c("tidyverse", "tm", "wordcloud", "tidytext", "ggplot2", "caret", "glmnet", "text2vec"))
library(tidyverse)
library(tm)
library(wordcloud)
library(tidytext)
library(ggplot2)
library(caret)
library(glmnet)
library(text2vec)

# ===========================
# 2. CHARGEMENT DES DONNÉES
# ===========================
# Charger le dataset (remplacer par votre chemin)
df <- read_csv("IMDB Dataset.csv")

# Afficher les premières lignes
glimpse(df)

# Vérifier les valeurs manquantes
sum(is.na(df))

# ===================================
# 3. ANALYSE EXPLORATOIRE (EDA)
# ===================================
# Distribution des sentiments
df %>%
  count(sentiment) %>%
  ggplot(aes(x = sentiment, y = n, fill = sentiment)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = c("positive" = "green", "negative" = "red")) +
  labs(title = "Distribution des sentiments", x = "Sentiment", y = "Nombre de critiques") +
  theme_minimal()

# Séparation des critiques positives et négatives
positive_reviews <- df %>% filter(sentiment == "positive") %>% pull(review)
negative_reviews <- df %>% filter(sentiment == "negative") %>% pull(review)

# Fonction pour générer un nuage de mots
generate_wordcloud <- function(text, title) {
  corpus <- Corpus(VectorSource(text))
  corpus <- tm_map(corpus, content_transformer(tolower))
  corpus <- tm_map(corpus, removePunctuation)
  corpus <- tm_map(corpus, removeNumbers)
  corpus <- tm_map(corpus, removeWords, stopwords("english"))
  
  wordcloud(corpus, 
            max.words = 100, 
            random.order = FALSE,
            colors = brewer.pal(8, "Dark2"),
            main = title)
}

# Nuage de mots pour les critiques positives
generate_wordcloud(positive_reviews, "Mots fréquents - Critiques POSITIVES")

# Nuage de mots pour les critiques négatives
generate_wordcloud(negative_reviews, "Mots fréquents - Critiques NÉGATIVES")

# Top 25 mots pour chaque sentiment
get_top_words <- function(text, n = 25) {
  text %>%
    tibble(text = .) %>%
    unnest_tokens(word, text) %>%
    anti_join(stop_words) %>%
    filter(!str_detect(word, "\\d")) %>%
    count(word, sort = TRUE) %>%
    head(n)
}

print("=== Top 25 mots POSITIFS ===")
get_top_words(positive_reviews)

print("=== Top 25 mots NÉGATIFS ===")
get_top_words(negative_reviews)

# ===================================
# 4. NETTOYAGE ET PRÉTRAITEMENT
# ===================================
clean_text <- function(text) {
  text %>%
    str_to_lower() %>%
    str_replace_all("<.*?>", "") %>% # Supprimer les balises HTML
    str_replace_all("[^[:alpha:]]", " ") %>% # Supprimer la ponctuation
    str_replace_all("\\s+", " ") %>% # Supprimer les espaces multiples
    str_trim() # Supprimer les espaces en début/fin
}

df <- df %>%
  mutate(cleaned_review = map_chr(review, clean_text))

# Tokenisation et lemmatisation (simplifiée)
df <- df %>%
  unnest_tokens(word, cleaned_review) %>%
  anti_join(stop_words) %>%
  mutate(word = SnowballC::wordStem(word)) %>% # Stemming
  group_by(sentiment, id = row_number()) %>%
  summarise(processed_review = paste(word, collapse = " ")) %>%
  ungroup()

# ===================================
# 5. EXTRACTION DES N-GRAMMES
# ===================================
get_top_ngrams <- function(text, n = 3, top_k = 10) {
  text %>%
    tibble(text = .) %>%
    unnest_tokens(ngram, text, token = "ngrams", n = n) %>%
    count(ngram, sort = TRUE) %>%
    head(top_k)
}

print("=== Top trigrammes POSITIFS ===")
get_top_ngrams(positive_reviews, 3)

print("=== Top 5-grammes POSITIFS ===")
get_top_ngrams(positive_reviews, 5)

print("=== Top trigrammes NÉGATIFS ===")
get_top_ngrams(negative_reviews, 3)

print("=== Top 5-grammes NÉGATIFS ===")
get_top_ngrams(negative_reviews, 5)

# ===================================
# 6. VECTORISATION ET MODÉLISATION
# ===================================
# Préparation des données
df_model <- df %>%
  mutate(label = ifelse(sentiment == "positive", 1, 0)) %>%
  select(processed_review, label)

# Création des partitions
set.seed(42)
train_index <- createDataPartition(df_model$label, p = 0.8, list = FALSE)
train_data <- df_model[train_index, ]
test_data <- df_model[-train_index, ]

# Vectorisation avec TF-IDF
prep_fun <- tolower
tok_fun <- word_tokenizer

it_train <- itoken(train_data$processed_review,
                   preprocessor = prep_fun,
                   tokenizer = tok_fun,
                   ids = train_data$id,
                   progressbar = FALSE)

vocab <- create_vocabulary(it_train)
vectorizer <- vocab_vectorizer(vocab)
dtm_train <- create_dtm(it_train, vectorizer)
tfidf <- Tfidf$new()
dtm_train_tfidf <- fit_transform(dtm_train, tfidf)

# Entraînement du modèle (Régression logistique avec régularisation)
model <- cv.glmnet(dtm_train_tfidf, 
                   train_data$label, 
                   family = "binomial",
                   type.measure = "auc",
                   nfolds = 5)

# Prétraitement des données de test
it_test <- itoken(test_data$processed_review,
                  preprocessor = prep_fun,
                  tokenizer = tok_fun,
                  ids = test_data$id,
                  progressbar = FALSE)

dtm_test <- create_dtm(it_test, vectorizer)
dtm_test_tfidf <- transform(dtm_test, tfidf)

# Prédictions
predictions <- predict(model, dtm_test_tfidf, type = "response", s = "lambda.min")
predicted_classes <- ifelse(predictions > 0.5, 1, 0)

# Évaluation
confusionMatrix(factor(predicted_classes), factor(test_data$label))

# ===================================
# 7. DÉPLOIEMENT ET TEST
# ===================================
predict_sentiment <- function(text) {
  cleaned_text <- clean_text(text)
  
  # Tokenisation et stemming
  tokens <- word_tokenizer(tolower(cleaned_text))
  tokens <- lapply(tokens, SnowballC::wordStem)
  processed_text <- paste(unlist(tokens), collapse = " ")
  
  # Vectorisation
  it_new <- itoken(processed_text, 
                   preprocessor = prep_fun,
                   tokenizer = tok_fun,
                   progressbar = FALSE)
  dtm_new <- create_dtm(it_new, vectorizer)
  dtm_new_tfidf <- transform(dtm_new, tfidf)
  
  # Prédiction
  prediction <- predict(model, dtm_new_tfidf, type = "response", s = "lambda.min")
  ifelse(prediction > 0.5, "POSITIVE", "NEGATIVE")
}

# Tests avec de nouvelles critiques
test_reviews <- c(
  "This movie was absolutely fantastic! The acting was superb and the plot was engaging.",
  "I hated this film. The story made no sense and the acting was terrible.",
  "It was okay, not great but not bad either.",
  "The director did an amazing job with this masterpiece!",
  "Waste of time and money. The worst movie I've seen this year."
)

print("=== Prédictions sur de nouvelles critiques ===")
for (i in seq_along(test_reviews)) {
  cat(sprintf("Review %d: %s\n", i, predict_sentiment(test_reviews[i])))
}

File: miniprojet.py
# -*- coding: utf-8 -*-
"""
Projet Text Mining - Analyse de Sentiments IMDB
Auteur : [Votre Nom]
Description : Classification des critiques de films en positives/négatives.
"""
# ======================
# 1. IMPORT DES LIBRAIRIES
# ======================
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from wordcloud import WordCloud
import re
from bs4 import BeautifulSoup
import nltk
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Téléchargement des ressources NLTK
nltk.download('stopwords')
nltk.download('punkt')
nltk.download('wordnet')

# ===========================
# 2. CHARGEMENT DES DONNÉES
# ===========================
# Charger le dataset (remplacer par votre chemin)
df = pd.read_csv("IMDB Dataset.csv")

# Afficher les premières lignes
print("\n=== Aperçu des données ===")
print(df.head())

# Vérifier les valeurs manquantes
print("\n=== Valeurs manquantes ===")
print(df.isnull().sum())

# ===================================
# 3. ANALYSE EXPLORATOIRE (EDA)
# ===================================
# Distribution des sentiments
print("\n=== Distribution des sentiments ===")
print(df['sentiment'].value_counts())

plt.figure(figsize=(6,4))
df['sentiment'].value_counts().plot(kind='bar', color=['green', 'red'])
plt.title("Distribution des sentiments")
plt.xticks(rotation=0)
plt.show()

# Séparation des critiques positives et négatives
positive_reviews = df[df['sentiment'] == 'positive']['review']
negative_reviews = df[df['sentiment'] == 'negative']['review']

# Fonction pour générer un nuage de mots
def generate_wordcloud(text, title):
    wordcloud = WordCloud(width=800, height=400, 
                          background_color='white',
                          max_words=100).generate(text)
    plt.figure(figsize=(10,5))
    plt.imshow(wordcloud, interpolation='bilinear')
    plt.axis('off')
    plt.title(title)
    plt.show()

# Nuage de mots pour les critiques positives
generate_wordcloud(" ".join(positive_reviews), "Mots fréquents - Critiques POSITIVES")

# Nuage de mots pour les critiques négatives
generate_wordcloud(" ".join(negative_reviews), "Mots fréquents - Critiques NÉGATIVES")

# Top 25 mots pour chaque sentiment
def get_top_words(reviews, n=25):
    words = " ".join(reviews).split()
    filtered_words = [word.lower() for word in words 
                     if word.lower() not in stopwords.words('english') 
                     and word.isalpha()]
    return Counter(filtered_words).most_common(n)

print("\n=== Top 25 mots POSITIFS ===")
print(get_top_words(positive_reviews))

print("\n=== Top 25 mots NÉGATIFS ===")
print(get_top_words(negative_reviews))

# ===================================
# 4. NETTOYAGE ET PRÉTRAITEMENT
# ===================================
def clean_text(text):
    # Supprimer les balises HTML
    text = BeautifulSoup(text, "html.parser").get_text()
    # Supprimer la ponctuation et caractères spéciaux
    text = re.sub(r'[^a-zA-Z\s]', '', text)
    # Convertir en minuscules
    text = text.lower()
    return text

df['cleaned_review'] = df['review'].apply(clean_text)

# Lemmatisation et suppression des stopwords
lemmatizer = WordNetLemmatizer()

def preprocess_text(text):
    tokens = word_tokenize(text)
    tokens = [lemmatizer.lemmatize(word) for word in tokens 
              if word not in stopwords.words('english')]
    return " ".join(tokens)

df['processed_review'] = df['cleaned_review'].apply(preprocess_text)

# ===================================
# 5. EXTRACTION DES N-GRAMMES
# ===================================
def get_top_ngrams(reviews, n=3, top_k=10):
    vec = CountVectorizer(ngram_range=(n, n)).fit(reviews)
    bag_of_words = vec.transform(reviews)
    sum_words = bag_of_words.sum(axis=0) 
    words_freq = [(word, sum_words[0, idx]) for word, idx in vec.vocabulary_.items()]
    return sorted(words_freq, key=lambda x: x[1], reverse=True)[:top_k]

print("\n=== Top trigrammes POSITIFS ===")
print(get_top_ngrams(positive_reviews, n=3))

print("\n=== Top 5-grammes POSITIFS ===")
print(get_top_ngrams(positive_reviews, n=5))

print("\n=== Top trigrammes NÉGATIFS ===")
print(get_top_ngrams(negative_reviews, n=3))

print("\n=== Top 5-grammes NÉGATIFS ===")
print(get_top_ngrams(negative_reviews, n=5))

# ===================================
# 6. VECTORISATION ET MODÉLISATION
# ===================================
# Division des données
X = df['processed_review']
y = df['sentiment'].map({'positive': 1, 'negative': 0})

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42)

# Vectorisation TF-IDF
tfidf = TfidfVectorizer(max_features=5000)
X_train_tfidf = tfidf.fit_transform(X_train)
X_test_tfidf = tfidf.transform(X_test)

# Entraînement du modèle
model = LogisticRegression(max_iter=1000)
model.fit(X_train_tfidf, y_train)

# Prédictions
y_pred = model.predict(X_test_tfidf)

# Évaluation
print("\n=== Performances du modèle ===")
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

# Matrice de confusion
cm = confusion_matrix(y_test, y_pred)
sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', 
            xticklabels=['Negative', 'Positive'], 
            yticklabels=['Negative', 'Positive'])
plt.title("Matrice de confusion")
plt.show()

# ===================================
# 7. DÉPLOIEMENT ET TEST
# ===================================
def predict_sentiment(text):
    cleaned_text = clean_text(text)
    processed_text = preprocess_text(cleaned_text)
    vectorized_text = tfidf.transform([processed_text])
    prediction = model.predict(vectorized_text)[0]
    return "POSITIVE" if prediction == 1 else "NEGATIVE"

# Tests avec de nouvelles critiques
test_reviews = [
    "This movie was absolutely fantastic! The acting was superb and the plot was engaging.",
    "I hated this film. The story made no sense and the acting was terrible.",
    "It was okay, not great but not bad either.",
    "The director did an amazing job with this masterpiece!",
    "Waste of time and money. The worst movie I've seen this year."
]

print("\n=== Prédictions sur de nouvelles critiques ===")
for i, review in enumerate(test_reviews, 1):
    print(f"Review {i}: {predict_sentiment(review)}")